kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java

```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
-                    try {
-                        if (bw != null) {
-                            bw.close();
-                        }
-                    } catch (IOException e) {
-                        s_logger.error("IOException on closing watchdog file", e);
-                    }
+                    performWatchdogActions();
+                } catch (KuraException e) {
+                    logger.error("Failed to perform watchdog actions.", e);
                 }
             }
+        } else {
+            if (this.configEnabled) {
+                refreshWatchdog();
+                this.isEnabled = true;
+            }
         }
     }
 
+    private synchronized void performWatchdogActions() throws KuraException {
+        // watchdogToStop is used to avoid multiple action executions and kill Kura after a while
+        if (!this.watchdogToStop) {
+            this.watchdogToStop = true;
+            runCommand("reboot");
+        } else {
+            refreshWatchdog();
+        }
+    }
+
+    private void runCommand(String command) throws KuraException {
+        try {
+            Process proc = Runtime.getRuntime().exec(command);
+            proc.waitFor();
+        } catch (IOException e) {
+            throw new KuraException(KuraErrorCode.OS_COMMAND_ERROR, e, "'" + command + "' failed");
+        } catch (InterruptedException e) {
+            throw new KuraException(KuraErrorCode.OS_COMMAND_ERROR, e, "'" + command + "' failed");
+        }
+    }
+
+    private void refreshWatchdog() {
+        writeWatchdogDevice("w");
+    }
+
     private void disableWatchdog() {
-        if (this.m_enabled) {
-            File f = null;
-            FileWriter bw = null;
+        if (this.isEnabled) {
+            writeWatchdogDevice("V");
+            this.isEnabled = false;
+        }
+    }
+
+    private synchronized void writeWatchdogDevice(String value) {
+        File f = null;
+        FileWriter bw = null;
+        try {
+            f = new File(this.watchdogDevice);
+            bw = new FileWriter(f);
+            bw.write(value);
+            logger.debug("write {} on watchdog device", value);
+        } catch (IOException e) {
+            logger.error("IOException on disable watchdog", e);
+        } finally {
             try {
-                f = new File(this.watchdogDevice);
-                bw = new FileWriter(f);
-                bw.write('V');
-                this.m_enabled = false;
-                s_logger.info("watchdog disabled");
+                if (bw != null) {
+                    bw.close();
+                }
             } catch (IOException e) {
-                s_logger.error("IOException on disable watchdog", e);
-            } finally {
+                logger.error("IOException on closing watchdog file", e);
+            }
+        }
+    }
+
+    private void readProperties(Map<String, Object> properties) {
+        if (properties.get(PROPERTY_PING_INTERVAL) != null) {
```

Comment:
If you don't rely on the properties also initialize explicitly fields to safe values (same as Metatype defaults IMHO)